zebra/zapi_msg.c

```diff
@@ -3268,6 +3269,171 @@ static inline void zread_rule(ZAPI_HANDLER_ARGS)
 	return;
 }
 
+static inline void zread_tc_qdisc(ZAPI_HANDLER_ARGS)
+{
+	struct zebra_tc_qdisc qdisc;
+	struct stream *s;
+	uint32_t total, i;
+
+	s = msg;
+	STREAM_GETL(s, total);
+
+	for (i = 0; i < total; i++) {
+		memset(&qdisc, 0, sizeof(qdisc));
+
+		qdisc.sock = client->sock;
+		STREAM_GETL(s, qdisc.qdisc.ifindex);
+		STREAM_GETL(s, qdisc.qdisc.kind);
+
+		if (hdr->command == ZEBRA_TC_QDISC_INSTALL)
+			zebra_tc_qdisc_install(&qdisc);
+		else
+			zebra_tc_qdisc_uninstall(&qdisc);
+	}
+
+stream_failure:
+	return;
+}
+
+static inline void zread_tc_class(ZAPI_HANDLER_ARGS)
+{
+	struct zebra_tc_class class;
+	struct stream *s;
+	uint32_t total, i;
+
+	s = msg;
+	STREAM_GETL(s, total);
+
+	for (i = 0; i < total; i++) {
+		memset(&class, 0, sizeof(class));
+
+		class.sock = client->sock;
+		STREAM_GETL(s, class.class.ifindex);
+		STREAM_GETL(s, class.class.handle);
+		STREAM_GETL(s, class.class.kind);
+		STREAM_GETQ(s, class.class.u.htb.rate);
+		STREAM_GETQ(s, class.class.u.htb.ceil);
+
+		if (hdr->command == ZEBRA_TC_CLASS_ADD)
+			zebra_tc_class_add(&class);
+		else
+			zebra_tc_class_delete(&class);
+	}
+
+stream_failure:
+	return;
+}
+
+static inline void zread_tc_filter(ZAPI_HANDLER_ARGS)
+{
+	struct zebra_tc_filter filter;
+	struct stream *s;
+	uint32_t total, i;
+
+	s = msg;
+	STREAM_GETL(s, total);
+
+	for (i = 0; i < total; i++) {
+		memset(&filter, 0, sizeof(filter));
+
+		filter.sock = client->sock;
+		STREAM_GETL(s, filter.filter.ifindex);
+		STREAM_GETL(s, filter.filter.handle);
+		STREAM_GETL(s, filter.filter.priority);
+		STREAM_GETL(s, filter.filter.protocol);
+		STREAM_GETL(s, filter.filter.kind);
+		switch (filter.filter.kind) {
+		case TC_FILTER_FLOWER: {
+			STREAM_GETL(s, filter.filter.u.flower.filter_bm);
+			uint32_t filter_bm = filter.filter.u.flower.filter_bm;
```

Comment:
I decided to put filter bitmask here which is different from [PBR](https://github.com/FRRouting/frr/pull/11908/commits/dbb438bbc53f64d8adbbc7f3dadd096731e3b4cb#diff-8e201155f5dddc9da6c8e814185d401c7395e5a3e165fe235f9e3e614741b155L3178). I think it will be better if we know the precise filter conditions from daemons instead of checking whether they are the default values?